torch/distributions/constraint_registry.py

```diff
@@ -0,0 +1,172 @@
+        """
+        Looks up a transform to constrained space, given a constraint object.
+        Usage::
+
+            constraint = Normal.params['scale']
+            scale = transform_to(constraint)(torch.zeros(1))  # constrained
+            u = transform_to(constraint).inv(scale)           # unconstrained
+
+        Args:
+            constraint (:class:`~torch.distributions.constraints.Constraint`):
+                A constraint object.
+
+        Returns:
+            A :class:`~torch.distributions.transforms.Transform` object.
+
+        Raises:
+            `NotImplementedError` if no transform has been registered.
+        """
+        # Look up by singleton instance.
+        try:
+            return self._registry[constraint]
+        except KeyError:
+            pass
+        # Look up by Constraint subclass.
+        try:
+            factory = self._registry[type(constraint)]
+        except KeyError:
+            raise NotImplementedError(
+                'Cannot transform {} constraints'.format(type(constraint).__name__))
+        return factory(constraint)
+
+
+biject_to = ConstraintRegistry()
+transform_to = ConstraintRegistry()
+
+################################################################################
+# Registration Table
+################################################################################
+
+biject_to.register(constraints.real, transforms.identity_transform)
+transform_to.register(constraints.real, transforms.identity_transform)
+
+biject_to.register(constraints.positive, transforms.ExpTransform())
+transform_to.register(constraints.positive, transforms.ExpTransform())
+
+
+@biject_to.register(constraints.greater_than)
+@transform_to.register(constraints.greater_than)
+def _transform_to_greater_than(constraint):
+    loc = -constraint.lower_bound
+    scale = loc.new([1]).expand_as(loc)
+    return transforms.ComposeTransform([transforms.AffineTransform(loc, scale),
+                                        transforms.ExpTransform()])
+
+
+@biject_to.register(constraints.less_than)
+@transform_to.register(constraints.less_than)
+def _transform_to_less_than(constraint):
+    loc = constraint.upper_bound
+    scale = loc.new([-1]).expand_as(loc)
+    return transforms.ComposeTransform([transforms.AffineTransform(loc, scale),
+                                        transforms.ExpTransform()])
+
+
+biject_to.register(constraints.unit_interval, transforms.SigmoidTransform())
+transform_to.register(constraints.unit_interval, transforms.SigmoidTransform())
+
+
+@biject_to.register(constraints.interval)
+@transform_to.register(constraints.interval)
+def _transform_to_interval(constraint):
+    span = constraint.upper_bound - constraint.lower_bound
+    loc = -span * constraint.lower_bound
+    scale = span.reciprocal()
+    return transforms.ComposeTransform([transforms.AffineTransform(loc, scale),
+                                        transforms.SigmoidTransform()])
+
+
```

Comment:
@dustinvtran This is the main difference between `transform_to()` and `biject_to()`. If you really need a bijection, then you should use `StickBreakingTransform` for `simplex`. If you don't really need a bijection, then `BoltzmannTransform` is cheaper, more numerically stable, and in theory works better with coordinate-wise optimizers like Adam.